### PR TITLE
Remove duplication in .mlis, expose MarshalError, add docs

### LINKED ISSUE #87: Cleanup mli files
- `idl.mli` and `rpc_lwt.mli` could be further simplified (removing unnecessarily exposed values) and reorganised a bit. To mention some:
  + the types in `Rpc_lwt.M` could all become abstract as the `_ fn` type
  + the `declare` function in the idl modules should return `'a res` 
  + some exceptions could be exposed to provide a better debugging experience

- `rpc_async.mli` is missing

When fixing these it would be nice to add some documentation

### PR DESCRIPTION
**Pull request by gaborigloi**
Fixes #87 

Signed-off-by: Gabor Igloi <gabor.igloi@citrix.com>

### CODE REVIEW

**Comment by gaborigloi** [comment]
We should also deduplicate the implementations themselves at one point, but that will be much harder.

**Comment by gaborigloi** [comment]
It would also be nice to upload the docs automatically somewhere.

**Comment by mseri** [comment]
I agree on both comments :)

File: src/lib/idl.mli [hunk 1]
```diff
@@ -180,35 +161,33 @@ type server_implementation
...
+    connect this rpc function to a real server that responds to client requests.
 
-  type _ fn
+		The implementations of each RPC method should be specified by passing it to
+		the corresponding function in the generated module.
 
-  val (@->) : 'a Param.t -> 'b fn -> ('a -> 'b) fn
-  val returning : 'a Param.t -> 'b Error.t -> ('a, 'b) comp fn
-  val declare : string -> string list -> 'a fn -> 'a res
+		Then we can the server itself can be obtained by passing the [implementation]
```
Comment (mseri) [inline src/lib/idl.mli]: Little typo: ‘Then the server itself’...

**Comment by mseri** [comment]
I think we could use the associated github pages for the documentation

**Comment by mseri** [comment]
Tomorrow I’ll play with dune-release and do a deserved upstream release and announce

**Comment by gaborigloi** [comment]
Thanks for the review, fixed the issues.
That sounds great, thanks!
I was wondering, should we split out more opam packages first, to avoid having a second breaking release afterwards? Because if we do #101 then we'll introduce the `angstrom` dependency (probably, if there isn't an ocamldoc parser already available), and then it'd be nice to add this extra dependency to a separate opam package, instead of making `rpclib` heavier.

**Comment by mseri** [comment]
I am still not sure ablut parsing ocamldoc by hands though

**Comment by mseri** [comment]
And I think it would be a minimal disruption anyways

**Comment by mseri** [comment]
I was starting to set-up the  auto push but I don’t have enough privileges to setup github pages. @jonludlam can you increase my privileges ro do it?

**Pull request merged**